src/gpu/kmeans/kmeans_h2o4gpu.cu

```diff
@@ -172,6 +178,456 @@ void random_centroids(int verbose, int seed, const char ord,
 "----------------------------------------------------------------------------\n"
 
 namespace h2o4gpukmeans {
+
+template<typename T>
+int kmeans_fit(int verbose, int seed, int gpu_idtry, int n_gputry,
+               size_t rows, size_t cols, const char ord,
+               int k, int max_iterations, int init_from_data,
+               int init_data, T threshold,
+               const T *srcdata, void **pred_centroids, void **pred_labels);
+
+template<typename T>
+int pick_point_idx_weighted(
+    thrust::host_device<T> data,
+    thrust::host_device<T> weights) {
+  T weight_sum = thrust::reduce(
+      weights.begin(),
+      weights.end()
+  );
+
+  T data_sum = 1.0;
+  if (NULL != data) {
+    data_sum = thrust::reduce(
+        data.begin(),
+        data.end()
+    );
+  }
+
+  // TODO should this be calculated for each record separately instead?
+  T r = ((T) rand() / (T) RAND_MAX) * weight_sum * data_sum;
+  int i = -1;
+  T cur_weight = 0.0;
+  T best_prob = 0.0;
+  int best_prob_idx = 0;
+  while (i < data.size() && cur_weight < r) {
+    T data_val = NULL == data ? 1.0 : data[i];
+
+    T curr_prob = data_val * weights[i];
+    if (curr_prob >= best_prob) {
+      best_prob = curr_prob;
+      best_prob_idx = i;
+    }
+
+    cur_weight += (data_val * weights[i]);
+    i += 1;
+  }
+  return i > -1 ? i : best_prob_idx;
+}
+
+/**
+ * Copies cols records, starting at position idx*cols from data to centroids. Removes them afterwards from data.
+ * Removes record from weights at position idx.
+ * @tparam T
+ * @param idx
+ * @param cols
+ * @param data
+ * @param weights
+ * @param centroids
+ */
+template<typename T>
+void add_centroid(int idx, int cols,
+                  thrust::host_device<T> &data,
+                  thrust::host_device<T> &weights,
+                  thrust::host_device<T> &centroids) {
+  for (int i = 0; i < cols; i++) {
```

Comment:
These kind of manipulations should be avoided on large vector types as they trigger resizing reallocating and copying. If the vectors are small it could be ok.

Reply:
yes, will benchmark it - the idea is that this whole method should only be called as the last step of kmeans||, where the dataset is small (only potential centroids).